File: dsa/week(4)/finaltouch.cpp
// 2. Run-time polymorphism

#include <iostream>
using namespace std;


class Base {
protected:
    int baseValue;

public:
    Base(int val) : baseValue(val) {}
    // Virtual function
    virtual void display() {
        cout << "Base class value: " << baseValue << endl;
    }
    // Pure virtual function
    virtual void show() = 0;
    friend void accessBaseValue(const Base& obj);
};

class Derived : public Base {
private:
    int derivedValue;
public:
    Derived(int baseVal, int derivedVal) : Base(baseVal), derivedValue(derivedVal) {}
    void display() override {
        cout << "Derived class values: BaseValue = " << baseValue << ", DerivedValue = " << derivedValue << endl;
    }
    void show() override {
        cout << "Derived class specific implementation of show()." << endl;
    }
    friend void accessDerivedValue(const Derived& obj);
};


void accessBaseValue(const Base& obj) {
    cout << "Accessing Base class protected value: " << obj.baseValue << endl;
}


void accessDerivedValue(const Derived& obj) {
    cout << "Accessing Derived class private value: " << obj.derivedValue << endl;
}

int main() {
    Derived obj(10, 20); 

    Base* basePtr = &obj;
    basePtr->display(); // Calls Derived's display() because it's virtual

    // Pure virtual function demonstration
    obj.show(); // Derived class must implement show()

    // Friend function demonstration
    accessBaseValue(obj);     // Access Base's protected member
    accessDerivedValue(obj);  // Access Derived's private member

    return 0;
}

File: dsa/week(4)/finaltouch.py
class Base:
    def __init__(self, base_val):
        self._base_value = base_val  # Protected member

    def display(self):
        print(f"Base class value: {self._base_value}")

    # Pure virtual function (enforced via raising NotImplementedError)
    def show(self):
        raise NotImplementedError("Subclasses must implement this method.")

    # Friend-like function to access Base class protected member
    @staticmethod
    def access_base_value(obj):
        print(f"Accessing Base class protected value: {obj._base_value}")


class Derived(Base):
    def __init__(self, base_val, derived_val):
        super().__init__(base_val)
        self.__derived_value = derived_val  
    def display(self):
        print(f"Derived class values: BaseValue = {self._base_value}, DerivedValue = {self.__derived_value}")

    
    def show(self):
        print("Derived class specific implementation of show().")

    # Friend-like function to access Derived class private member
    @staticmethod
    def access_derived_value(obj):
        print(f"Accessing Derived class private value: {obj.__derived_value}")


# Driver code
if __name__ == "__main__":
    obj = Derived(10, 20)

    # Virtual function demonstration
    base_ptr = obj
    base_ptr.display()  # Calls Derived's display() because Python supports dynamic dispatch

    # Pure virtual function demonstration
    obj.show()  # Derived class must implement show()

    # Friend-like function demonstration
    Base.access_base_value(obj)    # Access Base's protected member
    Derived.access_derived_value(obj)  # Access Derived's private member
